### PR TITLE
Reinforcement Learning: Deep Deterministic Policy Gradient 

### PR DESCRIPTION
**Pull request by tareknaser360**
# Description

This pull request implements the DDPG (Deep Deterministic Policy Gradient) algorithm, along with 2 test cases. 

## Implementation details
DDPG is an actor-critic algorithm designed for continuous action spaces. It combines deep neural networks with deterministic policy gradients to learn optimal policies in a continuous control setting.

Implemented four networks:
- `policyNetwork` (actor network)
- `targetPNetwork` (target actor network)
- `learningQNetwork` (critic network)
- `targetQNetwork` (target critic network)

# How Has This Been Tested?

- [x] Included a Pendulum test that successfully passes on my machine. The average reward achieved in the Pendulum environment is approximately -500, indicating successful learning.
- [x] Additionally, added a test for continuous action spaces, which also passes.

The test configurations for DDPG are adapted from the SAC (Soft Actor-Critic) implementation since both DDPG and SAC are policy gradient off-policy algorithms. This ensures consistent evaluation and comparison of the algorithms.

### CODE REVIEW

File: HISTORY.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 ### mlpack ?.?.?
 ###### ????-??-??
+  *  Reinforcement Learning: Deep Deterministic Policy Gradient (#3494).
```
Comment (zoq) [inline HISTORY.md:3]:
```suggestion
  * Reinforcement Learning: Deep Deterministic Policy Gradient (#3494).
```

Remove extra space.

Comment (tareknaser360) [reply HISTORY.md]: Done!
I also rebased on top of master.

File: src/mlpack/methods/reinforcement_learning/ddpg.hpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+ * @tparam UpdaterType How to apply gradients when training.
+ * @tparam ReplayType Experience replay method.
+ */
+template <
+  typename EnvironmentType,
+  typename QNetworkType,
+  typename PolicyNetworkType,
+  typename UpdaterType,
+  typename ReplayType = RandomReplay<EnvironmentType>
+>
```
Comment (shubham1206agra) [inline src/mlpack/methods/reinforcement_learning/ddpg.hpp:57]: tparam not matching

Comment (tareknaser360) [reply src/mlpack/methods/reinforcement_learning/ddpg.hpp]: Thanks for pointing this out.

File: src/mlpack/tests/q_learning_test.cpp [hunk 3]
```diff
@@ -565,3 +565,97 @@ TEST_CASE("SACForMultipleActions", "[QLearningTest]")
...
+}
+
+//! A test to ensure DDPG works with multiple actions in action space.
+TEST_CASE("DDPGForMultipleActions", "[QLearningTest]")
+{
+  FFN<EmptyLoss, GaussianInitialization>
+      policyNetwork(EmptyLoss(), GaussianInitialization(0, 0.1));
+  policyNetwork.Add(new Linear(128));
+  policyNetwork.Add(new ReLU());
+  policyNetwork.Add(new Linear(128));
```
Comment (zoq) [inline src/mlpack/tests/q_learning_test.cpp:623]: Do you think, we can reduce the size of the second linear layer, to make the test faster?

Comment (tareknaser360) [reply src/mlpack/tests/q_learning_test.cpp]: We can take out the second layer for the tests.
The networks used in the original PR are taken from the paper but we can always add an example later in the `examples` repository to demonstrate that.

For now, the modified networks converge as well. The average return for the Pendulum task is -970.

**Review by zoq — APPROVED** [review]
No more comments from my side, awesome work.
